ReactWindows/ReactNative/UIManager/UIViewOperationQueueCollection.cs

```diff
@@ -0,0 +1,542 @@
+                {
+                    queueInstance = new UIViewOperationQueueInstance(
+                        _reactContext, new NativeViewHierarchyManager(_viewManagerRegistry, rootViewDispatcher), reactChoreographer),
+                    rootViewCount = 1
+                };
+
+                // Add new tuple to map
+                _dispatcherToOperationQueueInfo.Add(rootViewDispatcher, queueInfo);
+
+                if (_active)
+                {
+                    // Simulate an OnResume from the correct dispatcher thread
+                    // (OnResume/OnSuspend/OnDestroy have this thread affinity, all other methods do enqueuings in a thread safe manner)
+                    DispatcherHelpers.RunOnDispatcher(rootViewDispatcher, queueInfo.queueInstance.OnResume, true); // inlining allowed
+                }
+            }
+            else
+            {
+                // Queue instance does exist.
+                // Increment the count of root views. This is helpful for the case the count reaches 0 so we can cleanup the queue.
+                queueInfo.rootViewCount++;
+            }
+
+            // Add tag
+            _reactTagToOperationQueue.Add(tag, queueInfo.queueInstance);
+
+            // Send forward
+            queueInfo.queueInstance.AddRootView(tag, rootView, themedRootContext);
+       }
+
+        /// <summary>
+        /// Enqueues an operation to remove the root view.
+        /// </summary>
+        /// <param name="rootViewTag">The root view tag.</param>
+        public void EnqueueRemoveRootView(int rootViewTag)
+        {
+            // Called on layout manager thread
+
+            UIViewOperationQueueInstance queue = GetQueueByTag(rootViewTag);
+
+            // Send forward
+            queue.EnqueueRemoveRootView(rootViewTag);
+
+            _reactTagToOperationQueue.Remove(rootViewTag);
+
+            // Do some maintenance/cleanup if needed.
+            // Find the queue info
+            var pair = _dispatcherToOperationQueueInfo.First(p => p.Value.queueInstance == queue);
+
+            // Decrement number of root views
+            pair.Value.rootViewCount--;
+
+            if (queue != MainUIViewOperationQueue)
+            {
+                if (pair.Value.rootViewCount == 0)
+                {
+                    // We can remove this queue and then destroy
+                    _dispatcherToOperationQueueInfo.Remove(pair.Key);
+
+                    // Simulate an OnDestroy from the correct dispatcher thread
+                    // (OnResume/OnSuspend/OnDestroy have this thread affinity, all other methods do enqueuings in a thread safe manner)
+                    DispatcherHelpers.RunOnDispatcher(pair.Key, queue.OnDestroy, true); // inlining allowed
+
+                    // Clean all the tags associated to this queue
+                    _reactTagToOperationQueue = _reactTagToOperationQueue.Where(p => p.Value != queue).ToDictionary(p => p.Key, p => p.Value);
```

Comment:
>_reactTagToOperationQueue = _reactTagToOperationQueue.Where(p => p.Value != queue).ToDictionary(p => p.Key, p => p.Value); [](start = 20, length = 122)

This is where I dream of "idle incremental scheduling" for those action queues...

Reply:
:+1:, it would would be great if we had a scheduler of some kind for our action queues. Can you file an issue and leave a comment here?

Alternatively (and this may not be the best approach), we can trade time for space and keep a Dictionary<RootNodeTag, HashSet<ViewTag>> and clean up using that.

---
In reply to: [166696023](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r166696023) [](ancestors = 166696023)

Reply:
HashSet it is (see below)

---
In reply to: [166925185](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r166925185) [](ancestors = 166925185,166696023)